crypto/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,438 @@
+        int exp_chunk_shift = exp_bit_no % 64;
+
+        /* Process 1-st exp window - just init result */
+        BN_ULONG red_table_idx_0 = expz[0][exp_chunk_no];
+        BN_ULONG red_table_idx_1 = expz[1][exp_chunk_no];
+        red_table_idx_0 = red_table_idx_0 >> exp_chunk_shift;
+        red_table_idx_1 = red_table_idx_1 >> exp_chunk_shift;
+
+        extract_multiplier_2x20_win5(red_Y[0], (const BN_ULONG*)red_table, (int)red_table_idx_0, 0);
+        extract_multiplier_2x20_win5(red_Y[1], (const BN_ULONG*)red_table, (int)red_table_idx_1, 1);
+
+        /* Process other exp windows */
+        for (exp_bit_no -= EXP_WIN_SIZE; exp_bit_no >= 0; exp_bit_no -= EXP_WIN_SIZE) {
+            /* Extract pre-computed multiplier from the table */
+            {
+                BN_ULONG T;
+                exp_chunk_no = exp_bit_no / 64;
+                exp_chunk_shift = exp_bit_no % 64;
+                {
+                    red_table_idx_0 = expz[0][exp_chunk_no];
+                    T = expz[0][exp_chunk_no + 1];
+
+                    red_table_idx_0 = red_table_idx_0 >> exp_chunk_shift;
+                    T = exp_chunk_shift == 0 ? 0 : T << (64 - exp_chunk_shift);
+                    red_table_idx_0 = (red_table_idx_0 ^ T) & table_idx_mask;
+
+                    extract_multiplier_2x20_win5(red_X[0], (const BN_ULONG*)red_table, (int)red_table_idx_0, 0);
+                }
+                {
+                    red_table_idx_1 = expz[1][exp_chunk_no];
+                    T = expz[1][exp_chunk_no + 1];
+
+                    red_table_idx_1 = red_table_idx_1 >> exp_chunk_shift;
+                    T = exp_chunk_shift == 0 ? 0 : T << (64 - exp_chunk_shift);
+                    red_table_idx_1 = (red_table_idx_1 ^ T) & table_idx_mask;
+
+                    extract_multiplier_2x20_win5(red_X[1], (const BN_ULONG*)red_table, (int)red_table_idx_1, 1);
+                }
+
+            }
+
+            /* Series of squaring */
+            DAMS((BN_ULONG*)red_Y, (const BN_ULONG*)red_Y, m, k0);
+            DAMS((BN_ULONG*)red_Y, (const BN_ULONG*)red_Y, m, k0);
+            DAMS((BN_ULONG*)red_Y, (const BN_ULONG*)red_Y, m, k0);
+            DAMS((BN_ULONG*)red_Y, (const BN_ULONG*)red_Y, m, k0);
+            DAMS((BN_ULONG*)red_Y, (const BN_ULONG*)red_Y, m, k0);
+
+            DAMM((BN_ULONG*)red_Y, (const BN_ULONG*)red_Y, (const BN_ULONG*)red_X, m, k0);
+        }
+    }
+
+    /* Convert result back in regular 2^52 domain */
+    memset(red_X, 0, sizeof(red_X));
+    red_X[0][0] = 1;
+    red_X[1][0] = 1;
+    DAMM(out, (const BN_ULONG*)red_Y, (const BN_ULONG*)red_X, m, k0);
+
+    /* Clear exponents */
+    OPENSSL_cleanse(expz, sizeof(expz));
+    OPENSSL_cleanse(red_Y, sizeof(red_Y));
+
+# undef DAMS
+# undef DAMM
+# undef EXP_DIGITS
+# undef RED_DIGITS
+# undef EXP_WIN_MASK
+# undef EXP_WIN_SIZE
+# undef BITSIZE_MODULUS
+}
+
+static ossl_inline uint64_t get_digit52(const uint8_t *in, int in_len)
+{
+    assert(NULL != in);
```

Comment:
Please use OPENSSL_assert() instead.

Reply:
OPENSSL_assert() macro generates code even with NDEBUG definition, which is not what we want to have in the critical path. I see a lot of usages of assert() in the OpenSSL, is this something that is not encouraged?

Reply:
We have a mixture of assert(), OPENSSL_assert() and ossl_assert(). It's currently unclear to me which one to use when, and we clearly need to document that somewhere. You can leave it as is.